node/src/types/block.rs

```diff
@@ -597,6 +601,32 @@ impl EraEnd {
     pub fn era_report(&self) -> &EraReport {
         &self.era_report
     }
+
+    pub fn hash(&self) -> Digest {
+        // Pattern match here leverages compiler to ensure every field is accounted for
+        let EraEnd {
+            next_era_validator_weights,
+            era_report,
+        } = self;
+        // Sort next era validator weights by descending weight. Assuming the top validators are
+        // online, a client can get away with just a few (plus a Merkle proof of the rest)
+        // to check finality signatures.
+        let descending_validator_weight_hashed_pairs: Vec<Digest> = next_era_validator_weights
+            .iter()
+            .sorted_by_key(|(_, weight)| **weight)
+            .rev()
+            .map(|(validator_id, weight)| {
+                let validator_hash =
+                    hash::hash(validator_id.to_bytes().expect("Could not hash validator"));
+                let weight_hash = hash::hash(weight.to_bytes().expect("Could not hash weight"));
+                hash::hash_pair(&validator_hash, &weight_hash)
+            })
+            .collect();
+        let hashed_next_era_validator_weights =
+            hash::hash_slice_rfold(&descending_validator_weight_hashed_pairs);
```

Comment:
A balanced tree would have the advantage that the average proof has length _O(log N)_ instead of _O(N/2)_.
And since the validators are sorted, a proof of an initial segment would still be short.

Reply:
Fixed in 8b08740f920054089076c3b868c8209829a22759

> And since the validators are sorted, a Merkle proof of an initial segment would still be short.

That's the idea.

Imagine you trust the switch block for era `N` and you want to validate the switch block for era `N+2`.  The idea is to just download and validate a Merkle proof of the top three validators or so (and the era number) for switch block `N+1`.  Using the top three you can then validate the `N+2` switch block.

It occurs to me that we'll need to extend the block body to have the total validator weight in order to do this trick.

I've created an issue in the backlog to capture this: https://github.com/casper-network/casper-node/issues/1622